File: binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int res = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        calculateDepth(root);
        return res;
        
    }
    public int calculateDepth(TreeNode node){
        if (node == null){
            return 0;
        }
        int left = calculateDepth(node.left);
        int right = calculateDepth(node.right);
        res = Math.max(res, left+right);
        return 1 + Math.max(left, right);
    }
}

File: depth-first-search/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        vector<int> res(1, 0);
        dfs(root, res);
        return res[0];

    }
public:
    int dfs(TreeNode* node, vector<int>& res){
        if(!node) return 0;
        int left = dfs(node->left, res);
        int right = dfs(node->right, res);
        
        res[0] = max(res[0], left+right);
        return 1 + max(left, right);

    }
};